# Secrets of The Dark Forest

- Автор: Бекишев С.В.
- Дата: 12.11.2025

### Цели и задачи проекта

<font color='#777778'>Перечислите цель проекта и задачи, которые вы решаете. Для этого можно использовать описание проекта, но полезнее будет сформулировать основную цель проекта самостоятельно.</font>

### Описание данных

<font color='#777778'>Здесь приведите описание данных.</font>

### Содержимое проекта

<font color='#777778'>Перечислите основные шаги проекта или напишите оглавление. Используйте описание проекта, чтобы зафиксировать основные шаги.</font>

---

## 1. Загрузка данных и знакомство с ними

- Загрузите необходимые библиотеки Python и данные датасета `/datasets/new_games.csv`.


In [7]:
import pandas as pd

In [8]:
df = pd.read_csv('new_games.csv')
df

Name Platform  Year of Release         Genre  \
0                         Wii Sports      Wii           2006.0        Sports   
1                  Super Mario Bros.      NES           1985.0      Platform   
2                     Mario Kart Wii      Wii           2008.0        Racing   
3                  Wii Sports Resort      Wii           2009.0        Sports   
4           Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing   
...                              ...      ...              ...           ...   
16951  Samurai Warriors: Sanada Maru      PS3           2016.0        Action   
16952               LMA Manager 2007     X360           2006.0        Sports   
16953        Haitaka no Psychedelica      PSV           2016.0     Adventure   
16954               Spirits & Spells      GBA           2003.0      Platform   
16955            Winning Post 8 2016      PSV           2016.0    Simulation   

       NA sales EU sales JP sales  Other sales  Critic Score User Score Rating  
0         41.36    28.96     3.77         8.45          76.0          8      E  
1         29.08     3.58     6.81         0.77           NaN        NaN    NaN  
2         15.68    12.76     3.79         3.29          82.0        8.3      E  
3         15.61    10.93     3.28         2.95          80.0          8      E  
4         11.27     8.89    10.22         1.00           NaN        NaN    NaN  
...         ...      ...      ...          ...           ...        ...    ...  
16951      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16952      0.00     0.01      0.0         0.00           NaN        NaN    NaN  
16953      0.00      0.0     0.01         0.00           NaN        NaN    NaN  
16954      0.01      0.0      0.0         0.00           NaN        NaN    NaN  
16955      0.00      0.0     0.01         0.00           NaN        NaN    NaN  

[16956 rows x 11 columns]

In [9]:
df.describe()

Year of Release      NA sales   Other sales  Critic Score
count     16681.000000  16956.000000  16956.000000   8242.000000
mean       2006.485522      0.262023      0.047087     68.926717
std           5.873102      0.808654      0.185577     13.944565
min        1980.000000      0.000000      0.000000     13.000000
25%        2003.000000      0.000000      0.000000     60.000000
50%        2007.000000      0.080000      0.010000     71.000000
75%        2010.000000      0.240000      0.030000     79.000000
max        2016.000000     41.360000     10.570000     98.000000

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


Комментарии по датафрейму:
- Существуют пропуски в столбцах:
  - Name
  - Year of Release
  - Genre
  - Critic Score: около половины значения NaN
  - User Score
  - Rating
- Типы данных:
  - Year of Release: сейчас тип float, но должен быть int
  - EU sales: сейчас тип object, но должен быть float
  - JP sales: сейчас тип object, но должен быть float
  - Critic Score: сейчас тип float, но можно сделать int
  - User Score: сейчас тип object, но можно убрать значения 'tbd', и преобразовать в float
- Названия столбцов не приведены к snake_case

---

## 2.  Проверка ошибок в данных и их предобработка


### 2.1. Названия, или метки, столбцов датафрейма

- Выведите на экран названия всех столбцов датафрейма и проверьте их стиль написания.
- Приведите все столбцы к стилю snake case. Названия должны быть в нижнем регистре, а вместо пробелов — подчёркивания.

In [13]:
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [14]:
# Преобразование названиq столбцов в snake_case
for column in df.columns:
    snake_column = column.lower().replace(' ', '_')
    df = df.rename(columns={column : snake_column})

In [15]:
# Проверка
df.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  eu_sales jp_sales  other_sales  critic_score user_score rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

### 2.2. Типы данных

- Если встречаются некорректные типы данных, предположите их причины.
- При необходимости проведите преобразование типов данных. Помните, что столбцы с числовыми данными и пропусками нельзя преобразовать к типу `int64`. Сначала вам понадобится обработать пропуски, а затем преобразовать типы данных.

In [17]:
# Создаем копию датафрейма для будущего сравнения с начальным
c_df = df.copy()

#### Проверка столбца 'name'

In [19]:
c_df[c_df['name'].isna() == True].head()

name platform  year_of_release genre  na_sales eu_sales jp_sales  \
661    NaN      GEN           1993.0   NaN      1.78     0.53      0.0   
14439  NaN      GEN           1993.0   NaN      0.00      0.0     0.03   

       other_sales  critic_score user_score rating  
661           0.08           NaN        NaN    NaN  
14439         0.00           NaN        NaN    NaN

In [20]:
c_df = c_df.dropna(subset=['name'])

Вывод: 
- Проверяем строки где name = NaN: 
- Найденные строки имеют малую смысловую нагрузку и их количество мало потому их можно удалить 

#### Проверка столбца platform

In [24]:
c_df['platform'].unique()

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

В столбце 'platform' никаких проблем не обнаружено.

#### Проеврка столбца year_of_release

In [27]:
# Строки с NaN в year_of_release
nan_year_of_release = c_df[c_df['year_of_release'].isna() == True]
nan_year_of_release.head()

name platform  year_of_release     genre  na_sales  \
183             Madden NFL 2004      PS2              NaN    Sports      4.26   
379            FIFA Soccer 2004      PS2              NaN    Sports      0.59   
458  LEGO Batman: The Videogame      Wii              NaN    Action      1.80   
477  wwe Smackdown vs. Raw 2006      PS2              NaN  Fighting      1.57   
611              Space Invaders     2600              NaN   Shooter      2.36   

    eu_sales jp_sales  other_sales  critic_score user_score rating  
183     0.26     0.01         0.71          94.0        8.5      E  
379     2.36     0.04         0.51          84.0        6.4      E  
458     0.97      0.0         0.29          74.0        7.9   E10+  
477     1.02      0.0         0.41           NaN        NaN    NaN  
611     0.14      0.0         0.03           NaN        NaN    NaN

In [183]:
print(f"Количество строк где year_of_release {len(c_df[c_df['year_of_release'].isna() == True])}")
print(f"Доля пустых year_of_release от датафрейма {round(len(nan_year_of_release) / len(df) * 100, 2)}%")

Количество строк где year_of_release 0
Доля пустых year_of_release от датафрейма 1.62%


In [30]:
# Для игр где year_of_release = NaN, но в названии есть приписка о годе, подстваляем этот год в year_of_release
# Подобная вставка может не отражать точный год, но скорее всего отклонение будет небольшое: +- 1 год

def check_year(row):
    try:
        year = int(row['name'][-4:])
        # Проверяем, что это год
        if year >= 1960 and year <= 2025:
            return year
        else:
            return row['year_of_release']
    except:
        return row['year_of_release']

c_df['year_of_release'] = c_df.apply(check_year, axis=1)
c_df.head()

name platform  year_of_release         genre  na_sales  \
0                Wii Sports      Wii           2006.0        Sports     41.36   
1         Super Mario Bros.      NES           1985.0      Platform     29.08   
2            Mario Kart Wii      Wii           2008.0        Racing     15.68   
3         Wii Sports Resort      Wii           2009.0        Sports     15.61   
4  Pokemon Red/Pokemon Blue       GB           1996.0  Role-Playing     11.27   

  eu_sales jp_sales  other_sales  critic_score user_score rating  
0    28.96     3.77         8.45          76.0          8      E  
1     3.58     6.81         0.77           NaN        NaN    NaN  
2    12.76     3.79         3.29          82.0        8.3      E  
3    10.93     3.28         2.95          80.0          8      E  
4     8.89    10.22         1.00           NaN        NaN    NaN

In [31]:
print(f"Количество строк с NaN в столбце year_of_release: {len(c_df[c_df['year_of_release'].isna() == True])}")

Количество строк с NaN в столбце year_of_release: 260


In [32]:
# Заполнение средним значением года по платформе для всех оставшихся year_of_release = NaN
def year_by_platform(row):
    if pd.isna(row['year_of_release']):
        group = c_df[c_df['platform'] == row['platform']]
        return round(group['year_of_release'].mean(), 0)
    else:
        return row['year_of_release']

In [33]:
c_df['year_of_release'] = c_df.apply(year_by_platform, axis=1)

# Меняем тип данных с float на int
c_df['year_of_release'] = c_df['year_of_release'].astype('int64')

# Проверка
c_df['year_of_release'].dtypes

dtype('int64')

Выводы:
- В изначальном датафрейме было 275 строк с year_of_release = NaN, что равно ~1.62% от общего кол-ва строк
- В некоторых строках было замечено, что в конце некоторых строк столбца 'name' содержится указание на год, такой год может не отражать точное значение, но часто такие указния имеют небольшое отклонение от реального года релиза. Таких строк нашлось немного - всего 15.
- Для всех остальных строк year_of_release = NaN было установлено среднее значение года релиза в зависимости от платформы

#### Проверка столбца genre

In [36]:
print(f"Количество NaN значений в столбце 'genre': {len(c_df[c_df['genre'].isna() == True])}\n")
print(c_df['genre'].unique())

Количество NaN значений в столбце 'genre': 0

['Sports' 'Platform' 'Racing' 'Role-Playing' 'Puzzle' 'Misc' 'Shooter'
 'Simulation' 'Action' 'Fighting' 'Adventure' 'Strategy' 'MISC'
 'ROLE-PLAYING' 'RACING' 'ACTION' 'SHOOTER' 'FIGHTING' 'SPORTS' 'PLATFORM'
 'ADVENTURE' 'SIMULATION' 'PUZZLE' 'STRATEGY']


In [37]:
c_df['genre'].unique()

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION', 'SHOOTER',
       'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE', 'SIMULATION',
       'PUZZLE', 'STRATEGY'], dtype=object)

В столбце 'genre' существуют повторяющиеся названия жанров в разных регисторов, приведем все к нижнему регистру

In [39]:
c_df['genre'] = c_df['genre'].str.lower()

#### Проверка столбца eu_sales

Так как в столбце 'na sales' количество ненулевых строк совпадает с количеством строк во всем датафрейме и тип данных установлен верно (float), предполагаем, что с этим столбцов все в порядке

In [42]:
# Пытаемся преобразовать к нужному, тем самым проверяя есть ли неподходящие значения
# c_df['eu_sales'] = c_df['eu_sales'].astype('float64')

In [43]:
# Выводим строки с "unknown"
print(f"Количество строк с 'unknown' = {len(c_df[c_df['eu_sales']=='unknown'])}")
c_df[c_df['eu_sales']=='unknown'].head()

Количество строк с 'unknown' = 6


name platform  year_of_release     genre  \
446                      Rhythm Heaven       DS             2008      misc   
802                        Dead Rising     X360             2006    action   
1131  Prince of Persia: Warrior Within      PS2             2004    action   
1132                         Far Cry 4     XOne             2014   shooter   
1394                   Sonic Advance 3      GBA             2004  platform   

      na_sales eu_sales jp_sales  other_sales  critic_score user_score rating  
446       0.55  unknown     1.93         0.13          83.0          9      E  
802       1.16  unknown     0.08         0.20          85.0        7.6      M  
1131      0.54  unknown      0.0         0.22          83.0        8.5      M  
1132      0.80  unknown     0.01         0.14          82.0        7.5      M  
1394      0.74  unknown     0.08         0.06          79.0        8.4      E

In [44]:
# Преобразуем значения 'unknown' в NaN
c_df['eu_sales'] = pd.to_numeric(c_df['eu_sales'], errors='coerce')

In [45]:
def eu_sales_filler(row):
    if pd.isna(row['eu_sales']):
        group = c_df[
            (c_df['platform'] == row['platform']) &
            (c_df['year_of_release'] == row['year_of_release']) &
            (c_df['genre'] == row['genre'])
        ]
        
        mean_sales = group['eu_sales'].mean()
        return mean_sales
    else:
        return row['eu_sales']

In [46]:
c_df['eu_sales'] = c_df.apply(eu_sales_filler, axis=1)

In [47]:
c_df['eu_sales'].dtype

dtype('float64')

#### Проверка столбца jp_sales

In [49]:
# Пытаемся преобразовать к нужному, тем самым проверяя есть ли неподходящие значения
# c_df['jp_sales'] = c_df['jp_sales'].astype('float64')

In [50]:
print(f"Количество строк с 'unknown' = {len(c_df[c_df['jp_sales']=='unknown'])}")
c_df[c_df['jp_sales']=='unknown'].head()

Количество строк с 'unknown' = 4


name platform  year_of_release  \
467                            Saints Row 2     X360             2008   
819                     UFC 2009 Undisputed     X360             2009   
1379                      Hello Kitty Party       DS             2007   
4732  Castlevania: The Dracula X Chronicles      PSP             2007   

         genre  na_sales  eu_sales jp_sales  other_sales  critic_score  \
467     action      1.94      0.79  unknown         0.28          81.0   
819   fighting      1.48      0.39  unknown         0.19          83.0   
1379      misc      0.78      0.51  unknown         0.12           NaN   
4732  platform      0.22      0.09  unknown         0.07          80.0   

     user_score rating  
467         8.1      M  
819         7.9      T  
1379        tbd      E  
4732        7.8      T

In [51]:
# Преобразуем значения 'unknown' в NaN
c_df['jp_sales'] = pd.to_numeric(c_df['jp_sales'], errors='coerce')

In [52]:
def jp_sales_filler(row):
    if pd.isna(row['jp_sales']):
        group = c_df[
            (c_df['platform'] == row['platform']) &
            (c_df['year_of_release'] == row['year_of_release']) &
            (c_df['genre'] == row['genre'])
        ]
        
        mean_sales = group['jp_sales'].mean()
        return mean_sales
    else:
        return row['jp_sales']

c_df['jp_sales'] = c_df.apply(jp_sales_filler, axis=1)

# Проверка
c_df['jp_sales'].dtype

#### В столбце other_sales нет пропущенных значений и установлен нужный тип данных, предполагаем, что здесь нет проблем

#### Проверка столбца critic_score

In [56]:
print(f"Количество NaN значений в столбце 'critic_score': {len(c_df[c_df['critic_score'].isna() == True])}\n")
c_df[c_df['critic_score'].isna() == True].head()

Количество NaN значений в столбце 'critic_score': 8712



name platform  year_of_release         genre  \
1          Super Mario Bros.      NES             1985      platform   
4   Pokemon Red/Pokemon Blue       GB             1996  role-playing   
5                     Tetris       GB             1989        puzzle   
9                  Duck Hunt      NES             1984       shooter   
10                Nintendogs       DS             2005    simulation   

    na_sales  eu_sales  jp_sales  other_sales  critic_score user_score rating  
1      29.08      3.58      6.81         0.77           NaN        NaN    NaN  
4      11.27      8.89     10.22         1.00           NaN        NaN    NaN  
5      23.20      2.26      4.22         0.58           NaN        NaN    NaN  
9      26.93      0.63      0.28         0.47           NaN        NaN    NaN  
10      9.05     10.95      1.93         2.74           NaN        NaN    NaN

Около половины строк датафрема являются строками с пустыми 'critic_score'. Удалять так много строк нельзя. Заполнение половины датафрейма на основе другой половины также кажется сомнительным, поэтому просто заменяем значения NaN на -1. Из df.desribe значем, что минимальное значение по столбцу равно 13, значит -1 подходит как значение-индикатор.

In [58]:
c_df['critic_score'] = c_df['critic_score'].fillna(-1)
c_df['critic_score'].dtype

dtype('float64')

#### Проверка столбца user_score

In [60]:
print(f"Количество NaN значений в столбце 'user_score': {len(c_df[c_df['user_score'].isna() == True])}\n")
c_df[c_df['user_score'].isna() == True].head()

Количество NaN значений в столбце 'user_score': 6802



name platform  year_of_release         genre  \
1          Super Mario Bros.      NES             1985      platform   
4   Pokemon Red/Pokemon Blue       GB             1996  role-playing   
5                     Tetris       GB             1989        puzzle   
9                  Duck Hunt      NES             1984       shooter   
10                Nintendogs       DS             2005    simulation   

    na_sales  eu_sales  jp_sales  other_sales  critic_score user_score rating  
1      29.08      3.58      6.81         0.77          -1.0        NaN    NaN  
4      11.27      8.89     10.22         1.00          -1.0        NaN    NaN  
5      23.20      2.26      4.22         0.58          -1.0        NaN    NaN  
9      26.93      0.63      0.28         0.47          -1.0        NaN    NaN  
10      9.05     10.95      1.93         2.74          -1.0        NaN    NaN

In [61]:
# Пытаемся преобразовать в float и заодно проверяем есть ли неподходящие значения
# c_df['user_score'] = c_df['user_score'].astype('float64')

In [62]:
print(f"Количество строк с 'tbd' в столбце user_score: {len(c_df[c_df['user_score']=='tbd'])}")
c_df[c_df['user_score']=='tbd'].head()

Количество строк с 'tbd' в столбце user_score: 2464


name platform  year_of_release  \
119                             Zumba Fitness      Wii             2010   
302            Namco Museum: 50th Anniversary      PS2             2005   
522                           Zumba Fitness 2      Wii             2011   
647                              uDraw Studio      Wii             2010   
659  Frogger's Adventures: Temple of the Frog      GBA             2003   

         genre  na_sales  eu_sales  jp_sales  other_sales  critic_score  \
119     sports      3.45      2.59       0.0         0.66          -1.0   
302       misc      2.08      1.35       0.0         0.54          61.0   
522     sports      1.51      1.03       0.0         0.27          -1.0   
647       misc      1.65      0.57       0.0         0.20          71.0   
659  adventure      2.15      0.18       0.0         0.07          73.0   

    user_score rating  
119        tbd      E  
302        tbd   E10+  
522        tbd      T  
647        tbd      E  
659        tbd      E

Кроме NaN в столбце user_score также есть значения 'tbd', значит для некоторых игр user_score еще уточняется. В данном случае, зная, что в данном столбце только положительные значения, заменить пустые строки и строки с 'tbd' на - 1.

In [64]:
# Замена NaN и 'tbd' на -1

c_df['user_score'] = c_df['user_score'].fillna(-1)

def tbd_replace(row):
    if row['user_score']=='tbd':
        return -1
    else:
        return row['user_score']

c_df['user_score'] = c_df.apply(tbd_replace, axis=1)

# Преобразуем в float и заодно проверяем что нет никаких проблем со значениями
c_df['user_score'] = c_df['user_score'].astype('float64')

#### Проверка столбца rating

In [66]:
c_df['rating'].unique()

array(['E', nan, 'M', 'T', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

Уже знаем, что в этом столбце есть пустые строки, теперь еще можем заметить, что существуют строки с устаревшим рейтингом 'K-A'. В 1998 году этот рейтинг был заменен на 'E'. Значит в датафрейме нужно изменить устаревший рейтинг на новый.

In [68]:
print(f"Количество строк с рейтингом 'K-A': {len(c_df[c_df['rating']=='K-A'])}")
c_df[c_df['rating']=='K-A']

Количество строк с рейтингом 'K-A': 3


name platform  year_of_release     genre  na_sales  \
658        Theme Hospital       PC             1997  strategy      2.30   
908    PaRappa The Rapper       PS             1996      misc      0.26   
16921             Worms 2       PC             1997  strategy      0.00   

       eu_sales  jp_sales  other_sales  critic_score  user_score rating  
658        0.10      0.00         0.00          -1.0         9.0    K-A  
908        0.16      1.46         0.03          92.0         7.4    K-A  
16921      0.01      0.00         0.00          -1.0         8.1    K-A

In [69]:
# Замена устаревшего рейтинга 'K-A' на 'E'
def rating_replace(row):
    if row['rating'] == 'K-A':
        return 'E'
    else:
        return row['rating']

c_df['rating'] = c_df.apply(rating_replace, axis=1)

In [152]:
print(f"Количество строк с пустыми значениями в 'rating': {len(c_df[c_df['rating'].isna() == True])}\n")
c_df[c_df['rating'].isna() == True].head()

Количество строк с пустыми значениями в 'rating': 0



Empty DataFrame
Columns: [name, platform, year_of_release, genre, na_sales, eu_sales, jp_sales, other_sales, critic_score, user_score, rating]
Index: []

Для того чтобы заменить пустые строки в rating применяем следующий алгоритм - для каждого жанра находим долю каждого рейтинга и принимаем, что в пустых строках сохранится такое же распределение т.е. в rating_prob_by_genre находятся вероятности рейтинга для каждого жанра

In [158]:
rating_prob_by_genre = {}
for genre in c_df['genre'].unique():
    rating_prob_by_genre[genre] = c_df[c_df['genre']==genre].groupby(by='rating')['genre'].count() / c_df[c_df['genre']==genre].groupby(by='rating')['genre'].count().sum()

import random

def get_rating(row):
    if pd.isna(row['rating']):
        cur_genre = row['genre']
        genre = rating_prob_by_genre[cur_genre].keys().to_list()
        weights = rating_prob_by_genre.get(cur_genre).to_list()
        res = random.choices(genre, weights=weights, k=1)
        return res[0]

    else:
        return row['rating']

c_df['rating'] = c_df.apply(get_rating, axis=1)

In [162]:
# Итоговая проврека что все сходится
c_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 16954 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16954 non-null  object 
 2   year_of_release  16954 non-null  int64  
 3   genre            16954 non-null  object 
 4   na_sales         16954 non-null  float64
 5   eu_sales         16954 non-null  float64
 6   jp_sales         16954 non-null  float64
 7   other_sales      16954 non-null  float64
 8   critic_score     16954 non-null  float64
 9   user_score       16954 non-null  float64
 10  rating           16954 non-null  object 
dtypes: float64(6), int64(1), object(4)
memory usage: 1.6+ MB
